core/contracts/Registry.sol

```diff
@@ -57,58 +61,103 @@ contract Registry is RegistryInterface, MultiRole {
         external
         onlyRoleHolder(uint(Roles.DerivativeCreator))
     {
-        // Create derivative pointer.
-        Pointer storage pointer = derivativePointers[derivativeAddress];
-
-        // Ensure that the pointer was not valid in the past (derivatives cannot be re-registered or double
-        // registered).
-        require(pointer.valid == PointerValidity.Invalid);
-        pointer.valid = PointerValidity.Valid;
+        Derivative storage derivative = addressToDerivatives[derivativeAddress];
+        require(addressToDerivatives[derivativeAddress].valid == DerivativeValidity.Invalid, "Can only register once");
 
+        // Store derivative address as a registered deriviative.
         registeredDerivatives.push(derivativeAddress);
 
         // No length check necessary because we should never hit (2^127 - 1) derivatives.
-        pointer.index = uint128(registeredDerivatives.length.sub(1));
+        derivative.index = uint128(registeredDerivatives.length.sub(1));
 
-        // Set up PartiesMap for this derivative.
-        PartiesMap storage partiesMap = derivativesToParties[derivativeAddress];
+        // For all parties in the array add them to the derivative party members.
+        // add the derivative as one of the party members own derivatives and store the index.
+        derivative.valid = DerivativeValidity.Valid;
         for (uint i = 0; i < parties.length; i = i.add(1)) {
-            partiesMap.parties[parties[i]] = true;
+            uint derivativeIndex = partyMembersToDerivatives[parties[i]].derivatives.push(derivativeAddress);
```

Comment:
nit: prefer to call this something like `newLength`, `length`, or something similar since the index is technically this value minus 1.

Reply:
good point. Will update.